File: parent/3310 -- Remove Methods From Project/mine2.go
package main

func remainingMethods(n int, k int, invocations [][]int) []int {
	adjList := map[int][]int{}
	infected := map[int]bool{}
	var res []int

	for _, invo := range invocations {
		node, nei := invo[0], invo[1]
		adjList[node] = append(adjList[node], nei)
	}

	var dfs_infect func(node int)
	dfs_infect = func(node int) {
		infected[node] = true

		for _, nei := range adjList[node] {
			if _, exists := infected[nei]; !exists {
				dfs_infect(nei)
			}
		}
	}

	dfs_infect(k)

	for _, invo := range invocations {
		node, nei := invo[0], invo[1]
		_, nodeExists := infected[node]
		_, neiExists := infected[nei]
		if !nodeExists && neiExists {
			for i := range n {
				res = append(res, i)
			}
			return res
		}

	}
	for i := range n {
		if _, exists := infected[i]; !exists {
			res = append(res, i)
		}
	}
	return res
}

File: parent/3310 -- Remove Methods From Project/mine2.py
from collections import defaultdict

class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        # More optimal for time and space
        # Time: O(n+m)
        # Space: O(n+m)
                
        adj_list = defaultdict(list)
        for node, nei in invocations:
            adj_list[node].append(nei)
        
        infected = set()

        def dfs_infect(node):
            infected.add(node)

            for nei in adj_list[node]:
                if nei not in infected:
                    dfs_infect(nei)
        
        dfs_infect(k)
        # Slight optimization: No need for second dfs
        # Just check if any uninfected node directly invokes an infected node
        for u,v in invocations:
            if u not in infected and v in infected:
                return list(range(n))
        
        return [x for x in range(n) if x not in infected]